js/src/components/SEMrushRelatedKeyphrasesModal.js

```diff
@@ -79,8 +88,60 @@ class SEMrushRelatedKeyphrasesModal extends Component {
 			"status=0",
 		];
 
-		window.open( url, "SEMrush_login", features.join( "," ) );
-		e.preventDefault();
+		if ( ! this.popup || this.popup.closed ) {
+			this.popup = window.open( url, "SEMrush_login", features.join( "," ) );
+		}
+		this.popup.focus();
+		window.addEventListener( "message", this.listenToMessages, false );
+	}
+
+	/**
+	 * Listens to message events from the SEMrush popup.
+	 *
+	 * @param {event} event The message event.
+	 *
+	 * @returns {void}
+	 */
+	listenToMessages( event ) {
+		const object = event.data;
+		const source = event.source;
+		const origin = event.origin;
+
+		// Check that the message comes from the expected origin.
+		if ( origin !== "https://oauth.semrush.com" || this.popup !== source ) {
+			return;
+		}
+
+		if ( object.type === "semrush:oauth:success" ) {
+			this.popup.close();
+			// Stop listening to messages, since the popup is closed.
+			window.removeEventListener( "message", this.listenToMessages, false );
+
+			try {
+				const url = new URL( object.url );
+				const code = url.searchParams.get( "code" );
+				apiFetch( {
+					path: "yoast/v1/semrush/authenticate",
+					method: "POST",
+					data: { code: code },
+				} ).then( res => {
```

Comment:
Consider using `async/await` to make these promises easier to read. The SEMrush request to retrieve data uses this approach.

Reply:
Done